mrmustard/physics/wires.py

```diff
@@ -169,242 +310,255 @@ def __init__(
+        for i, m in enumerate(sorted(classical_in)):
+            self.classical_wires.add(ClassicalWire(mode=m, is_out=False, index=n + i))
+
+    @classmethod
+    def from_wires(
+        cls,
+        quantum: Iterable[QuantumWire] = (),
+        classical: Iterable[ClassicalWire] = (),
+        copy: bool = False,
+    ) -> Wires:
+        r"""
+        Returns a new Wires object with references to the given wires.
+        If copy is True, the wires are copied, otherwise they are referenced.
+        """
+        w = Wires()
+        w.quantum_wires = set(quantum) if not copy else {q.copy() for q in quantum}
+        w.classical_wires = set(classical) if not copy else {c.copy() for c in classical}
+        return w
+
+    def copy(self, new_ids: bool = False) -> Wires:
+        """Returns a deep copy of this Wires object."""
+        return Wires.from_wires(
+            quantum={q.copy(new_ids) for q in self.quantum_wires},
+            classical={c.copy(new_ids) for c in self.classical_wires},
         )
-        self._len = None
 
-        # The "parent" wires object, if any. This is ``None`` for freshly initialized
-        # wires objects, and ``not None`` for subsets.
-        self._original = None
-
-        # Adds elements to the cache when calling ``__getitem__``
-        self._mode_cache = {}
+    ###### NEW WIRES ######
 
     @cached_property
     def adjoint(self) -> Wires:
         r"""
-        New ``Wires`` object obtained by swapping ket and bra wires.
+        New ``Wires`` object with the adjoint quantum wires (ket becomes bra and vice versa).
         """
         return Wires(
-            self.args[2],
-            self.args[3],
-            self.args[0],
-            self.args[1],
-            self.args[4],
-            self.args[5],
+            modes_out_bra=self.output.ket.modes,
+            modes_in_bra=self.input.ket.modes,
+            modes_out_ket=self.output.bra.modes,
+            modes_in_ket=self.input.bra.modes,
+            classical_out=self.output.classical.modes,
+            classical_in=self.input.classical.modes,
         )
 
     @cached_property
-    def bra(self) -> Wires:
+    def dual(self) -> Wires:
         r"""
-        New ``Wires`` object with only bra wires.
+        New ``Wires`` object with dual quantum and classical wires (input becomes output and vice versa).
         """
-        ret = Wires(modes_out_bra=self.args[0], modes_in_bra=self.args[1])
-        ret._original = self.original or self
-        return ret
+        return Wires(
+            modes_out_bra=self.input.bra.modes,
+            modes_in_bra=self.output.bra.modes,
+            modes_out_ket=self.input.ket.modes,
+            modes_in_ket=self.output.ket.modes,
+            classical_out=self.input.classical.modes,
+            classical_in=self.output.classical.modes,
+        )
 
-    @cached_property
-    def classical(self) -> Wires:
+    def to_fock(self, shape: tuple[int, ...]) -> Wires:
```

Comment:
A point in favour of keeping the representation information in `Representation` rather that `Wires`: with this we now need to initialize a new instance of `Wires` every time we do a representation change. Instead, by separating out the responsibility into the `Representation` class a representation change is simply updating the value in a dictionary. For this reason (as well as conceptually) I think we should keep that logic in the `Representation` (and maybe even introduce a dataclass to wrap `repr`, `repr_params`)

Reply:
I agree that it is not necessary to pass the shape for Fock, but I want to push back on where we put the information. I hear you, that we need to rebuild the wires, but that is hardly the bottleneck given all the other things we do in a representation change. And the wires are part of the representation anyway, so afaik it makes more sense to centralize the wire information into the wires. I'll review the way I pass this info into the wires though.